Ship_info vectorization, ship templates.

MageKing17 opened the pull request:
Revamps ship_info with constructors, assignment operators, and a destructor, turns Ship_info into a vector, and reimplements ship templates. See also [Mantis 3083](http://scp.indiegames.us/mantis/view.php?id=3083).

MageKing17 commented:
Ah, yeah, that would probably be the move constructor and move-assignment operator. I couldn't figure out how to make MSVC actually use them when resizing the vector ("noexcept" isn't supported until MSVS 2015), but maybe it was due to the lack of some of the changes made in the C++11 branch?

Given that I used C++11 stuff for this branch, the C++11 stuff should probably get merged first.

Echelon9 commented:
Probably goes without saying, but alongside implementing any Big Three (constructor, copy, destructor) it would be highly useful to ensure this PR doesn't throw any new AddressSanitizer errors.

I found that essential as part of my testing converting the jump node code into a proper CJumpNode class.

chief1983 commented:
Is this waiting on PR #50 too?

chief1983 commented:
Or maybe #52 is what I meant

chief1983 commented:
I was getting errors about cend and whatnot, which is why I asked.  I added the c++11 settings from #52 into this ticket, and tried to recompile.  Now I'm getting another error though.

https://gist.github.com/chief1983/bd4d2a49b2430797b6b7

MageKing17 commented:
I guess it's kind of waiting on both.

Try adding `#include <algorithm>` to ship.cpp and see if it compiles then?

chief1983 commented:
FFS...
```
CompileC build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/redalert.o /Users/cliff.gordon/fs2_open_git/code/missionui/redalert.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
    cd /Users/cliff.gordon/fs2_open_git/projects/Xcode4
    export LANG=en_US.US-ASCII
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c++ -arch x86_64 -fmessage-length=216 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -fcolor-diagnostics -std=c++11 -stdlib=libc++ -Wno-trigraphs -fpascal-strings -O0 -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-non-virtual-dtor -Wno-overloaded-virtual -Wno-exit-time-destructors -Wno-missing-braces -Wparentheses -Wswitch -Wunused-function -Wno-unused-label -Wno-unused-parameter -Wunused-variable -Wunused-value -Wempty-body -Wuninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wconstant-conversion -Wint-conversion -Wbool-conversion -Wenum-conversion -Wshorten-64-to-32 -Wno-newline-eof -Wno-c++11-extensions -D_DEBUG -DSCP_UNIX -DUSE_OPENAL -DAPPLE_APP -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk -fasm-blocks -fstrict-aliasing -Wdeprecated-declarations -Winvalid-offsetof -mmacosx-version-min=10.7 -g -Wno-sign-conversion -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/code.hmap -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/Debug/include -I../../code -IFrameworks/Ogg.framework/Headers -IFrameworks/Vorbis.framework/Headers -IFrameworks/Theora.framework/Headers -IFrameworks/SDL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk/System/Library/Frameworks/OpenAL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk/System/Library/Frameworks/OpenGL.framework/Headers -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/DerivedSources/x86_64 -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/DerivedSources -Wno-write-strings -Wno-char-subscripts -Wno-shorten-64-to-32 -F/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/Debug -FFrameworks -MMD -MT dependencies -MF /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/redalert.d --serialize-diagnostics /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/redalert.dia -c /Users/cliff.gordon/fs2_open_git/code/missionui/redalert.cpp -o /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/redalert.o
In file included from /Users/cliff.gordon/fs2_open_git/code/missionui/redalert.cpp:12:
In file included from ../../code/missionui/redalert.h:15:
In file included from ../../code/globalincs/pstypes.h:645:
In file included from ../../code/globalincs/vmallocator.h:12:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/c++/v1/unordered_map:724:5: error: static_assert failed "Invalid allocator::value_type"
    static_assert((is_same<value_type, typename allocator_type::value_type>::value),
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /Users/cliff.gordon/fs2_open_git/code/missionui/redalert.cpp:12:
In file included from ../../code/missionui/redalert.h:15:
In file included from ../../code/globalincs/pstypes.h:645:
../../code/globalincs/vmallocator.h:131:34: note: in instantiation of template class 'std::__1::unordered_map<int, int, std::__1::hash<int>, std::__1::equal_to<int>, SCP_vm_allocator<int> >' requested here
class SCP_unordered_map : public std::unordered_map< Key, T, std::hash<Key>, std::equal_to<Key>, SCP_vm_allocator< T > > { };
                                 ^
/Users/cliff.gordon/fs2_open_git/code/missionui/redalert.cpp:825:30: note: in instantiation of template class 'SCP_unordered_map<int, int>' requested here
        SCP_unordered_map<int, int> Wing_pobjects_deleted;
                                    ^
/Users/cliff.gordon/fs2_open_git/code/missionui/redalert.cpp:826:31: error: no type named 'iterator' in 'SCP_unordered_map<int, int>'
        SCP_unordered_map<int, int>::iterator ii;
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
2 errors generated.
```
After merging all three pull requests I got that error now.

chief1983 commented:
That's related to something in PR #52 though it seems.

chief1983 commented:
I should also probably mention that I have always merged master in when doing compile tests on these PRs.  Maybe that's a bad practice though, but I like to know how they'll behave when merged in as opposed to when pushed.

chief1983 commented:
Ok, since compilation of #52 on OS X was fixed I merged it all back together and still get the original swap issue, even after adding #include <algorithm> to ship.cpp again.

MageKing17 commented:
Looks like a bug in clang: https://llvm.org/bugs/show_bug.cgi?id=22106

chief1983 commented:
Well, then maybe it's already fixed in Xcode 6, 8nm on 5.1.1 on 10.8 until i setup my new laptop.  But is there any workaround for those compiling on mountain lion stuck on xcode 5?

MageKing17 commented:
Yeah; can just do a memcpy() instead. I'm already not trying to completely swap the two objects; just the things relevant to the destructor. I only swapped instead of memcpy()-ing the arrays because it seemed "more c++11".

MageKing17 commented:
I've merged it with latest master, updated to add the new member (and correct some typos in the old code), and changed the multidimensional array swaps to use memcpy() instead.

chief1983 commented:
Hmm, not sure what happened now, re-merged 41, 50, and 52, and got this now:
```
CompileC build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/aicode.o /Users/cliff.gordon/fs2_open_git/code/ai/aicode.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
    cd /Users/cliff.gordon/fs2_open_git/projects/Xcode4
    export LANG=en_US.US-ASCII
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c++ -arch x86_64 -fmessage-length=216 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -fcolor-diagnostics -std=c++11 -stdlib=libc++ -Wno-trigraphs -fpascal-strings -O0 -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-non-virtual-dtor -Wno-overloaded-virtual -Wno-exit-time-destructors -Wno-missing-braces -Wparentheses -Wswitch -Wunused-function -Wno-unused-label -Wno-unused-parameter -Wunused-variable -Wunused-value -Wempty-body -Wuninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wconstant-conversion -Wint-conversion -Wbool-conversion -Wenum-conversion -Wshorten-64-to-32 -Wno-newline-eof -Wno-c++11-extensions -D_DEBUG -DSCP_UNIX -DUSE_OPENAL -DAPPLE_APP -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk -fasm-blocks -fstrict-aliasing -Wdeprecated-declarations -Winvalid-offsetof -mmacosx-version-min=10.7 -g -Wno-sign-conversion -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/code.hmap -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/Debug/include -I../../code -IFrameworks/Ogg.framework/Headers -IFrameworks/Vorbis.framework/Headers -IFrameworks/Theora.framework/Headers -IFrameworks/SDL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk/System/Library/Frameworks/OpenAL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk/System/Library/Frameworks/OpenGL.framework/Headers -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/DerivedSources/x86_64 -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/DerivedSources -Wno-write-strings -Wno-char-subscripts -Wno-shorten-64-to-32 -F/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/Debug -FFrameworks -MMD -MT dependencies -MF /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/aicode.d --serialize-diagnostics /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/aicode.dia -c /Users/cliff.gordon/fs2_open_git/code/ai/aicode.cpp -o /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/aicode.o
In file included from /Users/cliff.gordon/fs2_open_git/code/ai/aicode.cpp:25:
../../code/ship/ship.h:1438:2: error: unknown type name 'SCP_hash_map'
        SCP_hash_map<int, void*> glowpoint_bank_override_map;
        ^
../../code/ship/ship.h:1438:14: error: expected member name or ';' after declaration specifiers
        SCP_hash_map<int, void*> glowpoint_bank_override_map;
        ~~~~~~~~~~~~^
2 errors generated
```
Seems to happen for each file that includes ship.h.  I don't see anything related to your commits in particular that would have caused that though, off the top of my head, since none seemed to change ship.h.

asarium commented:
I changed `SCP_hash_map` to `SCP_unordered_map` in one PR. That's probably causing it.

MageKing17 commented:
glowpoint_bank_override_map was added by the deferred lighting code merge (which was after PR #52's commit related to SCP_unordered_map). In other words, that particular problem has nothing to do with *this* pull request. ;P

chief1983 commented:
So, with master merged into that PR, it won't compile and needs to be updated now I take it?

niffiwan commented:
I also found that in the F3 lab all the ships are using the GTF Ares model. The techroom displays the correct models, and the F3 lab also works on current master.

MageKing17 commented:
Everything showing the GTF Ares model? That sounds like a similar problem I had with the non-F3 lab where everything showed the GTT Elysium model; let me see if it's a similar cause.

MageKing17 commented:
Ignore the second paragraph of the merge commit; I thought I had those changes staged at the time, but SourceTree apparently ignored that fact; not sure why.

chief1983 commented:
This isn't waiting on any other fixes or blocks is it?  If not, let's get this merged.

DahBlount commented:
There isn't anything that pops out at me. Build was successful on all configurations using MSVC 2013.
If anyone else thinks this should be merged, then by all means do so.

MageKing17 commented:
This PR affects such a wide swath of code that I'd really like at least one other person to review it (not just double-checking the code, but also running it and making sure it doesn't screw something up) before this gets merged.

chief1983 commented:
We could do a Test Builds post perhaps.

chief1983 commented:
Test builds posted http://www.hard-light.net/forums/index.php?topic=89942.0

Goober5000 commented:
MageKing17, there are probably places where ship_info pointers are stored which reference locations in the ship_info vector.  These pointers are valid for only as long as the vector stays the same size, because if adding an element causes a need for more memory, it is possible for the entire vector to be moved elsewhere and for every pointer to become invalid.

It will be necessary to ensure that ship_info pointers are only used after parsing is complete (because I'm pretty sure parsing is the only opportunity for the size to change).  I think most if not all objects in the code use ship_info_index instead of a ship_info pointer, and the only occurrences of ship_info* are in function parameters, but this should be double-checked.

In particular, if any global ship_info pointers are maintained *during* parsing, then this is dangerous, because it is theoretically possible for the vector to change memory locations multiple times before parsing is finished.

MageKing17 commented:
The only instance of a non-temporary `ship_info*` pointer I could find was on the `WarpEffect` object, which is created by `ship_set_warp_effects()`, which *definitely* shouldn't be being called during table parsing.

z64555 commented on code/stats/scoring.cpp:
```diff
@@ -300,8 +300,9 @@ void scoring_eval_badges(scoring_struct *sc)
 	total_kills = 0;
-	for (i = 0; i < MAX_SHIP_CLASSES; i++ ) {
-		if ( (Ship_info[i].flags & SIF_FIGHTER) || (Ship_info[i].flags & SIF_BOMBER) ) {
+	for (auto it = Ship_info.cbegin(); it != Ship_info.cend(); ++it ) {
```
Any special reason for using an iterator here instead of int indexes?

z64555 commented on code/ship/ship.cpp:
```diff
@@ -16442,12 +17100,10 @@ int ship_get_reinforcement_team(int r_index)
-	for(idx=0; idx<Num_ship_classes; idx++){
-		if((Ship_info[idx].model_num >= 0) && model_find_texture(Ship_info[idx].model_num, bitmap) == 1){
-			return idx;
+	for (auto it = Ship_info.cbegin(); it != Ship_info.cend(); ++it) {
```
Iterators vs. int indexes?

z64555 commented on code/ship/ship.cpp:
```diff
@@ -15808,7 +16466,7 @@ void ship_page_in()
 					sip->subsystems[j].model_num = -1;
 
-				ship_copy_subsystem_fixup(sip);
+				ship_copy_subsystem_fixup(&(*sip));
```
workaround?

MageKing17 replied:
`ship_copy_subsystem_fixup()` takes a pointer. `sip` is now an iterator. To get a pointer to the `ship_info` object that the iterator points to, we need to deference the iterator and then get the address of that object. This ugly piece of code is the result. If there's a cleaner way to do it, I don't know one, and `making ship_copy_subsystem_fixup()` take an iterator instead of a pointer would require changing the other places it gets called to use iterators instead of pointers (which would be even more hackish).

Goober5000 replied:
This is the correct idiom.  It appears elsewhere in our codebase and I've seen it used in other places online as well.

asarium commented on code/io/keycontrol.cpp:
```diff
@@ -490,11 +490,11 @@ void debug_cycle_player_ship(int delta)
 		}
 		if ( si_index < 0 ){
-			si_index = Num_ship_classes - 1;
+			si_index = Ship_info.size() - 1;
```
I think this is missing a `static_cast<int>(...)`.

asarium commented on code/io/keycontrol.cpp:
```diff
@@ -538,10 +538,10 @@ void debug_cycle_targeted_ship(int delta)
 			si_index = 0;
 		if ( si_index < 0 )
-			si_index = Num_ship_classes-1;
+			si_index = Ship_info.size() - 1;
```
Same here.

asarium commented on code/menuui/techmenu.cpp:
```diff
@@ -801,7 +801,7 @@ void techroom_change_tab(int num)
 
 				// make sure that at least the default entry is cleared out if we didn't grab anything
-				if (Num_ship_classes && !Ship_list_size) {
+				if (Ship_info.size() && !Ship_list_size) {
```
Using `!Ship_info.empty()` would be easier to read here.

asarium commented on code/mission/missiongoals.cpp:
```diff
@@ -978,10 +978,11 @@ void mission_process_event( int event )
-	if (result && (Mission_events[event].trigger_count > 0) && (Mission_events[event].flags & MEF_USING_TRIGGER_COUNT) ) {
-		Mission_events[event].trigger_count--;
+	if (result && (Mission_events[event].trigger_count != 0) && (Mission_events[event].flags & MEF_USING_TRIGGER_COUNT) ) {
+		if (Mission_events[event].trigger_count > 0)
```
How is this change related to ship_info vectorization? There is another one below that also doesn't look like it's related.

MageKing17 replied:
Oh wow, whoops. I had an uncommitted change there that I thought I'd stashed; no idea how it accidentally wound up in this branch. Looks like it slipped in with bc0c7cf (the original vectorization commit), which was busy enough that I can imagine having accidentally included something unrelated.

asarium commented on code/missionui/missionscreencommon.cpp:
```diff
@@ -1081,11 +1081,11 @@ void wss_maybe_restore_loadout()
 
 			for ( j = 0; j < MAX_SHIP_WEAPONS; j++ ) {
-				if ((slot->wep[j] >= 0) && (slot->wep[j] < MAX_WEAPON_TYPES)) {
+				if ((slot->wep[j] >= 0) && (slot->wep[j] < Num_weapon_types)) {
```
Even though this looks right I think it would be better to have this in another PR. (there are more similar changes below)

MageKing17 replied:
This was done in commit 3a7307d and the reason it's here is because it's not completely unrelated; I suppose it could be reverted and recommitted in a separate branch if we *really* want strictly-defined limitations on this PR, but that'll clutter the history even more than it already is...

asarium replied:
Ok, it just looked a bit unrelated but if it's related then it's ok (you can clean up your history by using interactive rebase).

asarium commented on code/network/multiutil.cpp:
```diff
@@ -757,16 +751,16 @@ int multi_create_player( int net_player_num, player *pl, const char* name, net_a
 		}
 
-		if (i == Num_ship_classes)
+		if (player_ship_class == -1)	// if ship_class is -1 (as the condition to enter this branch at all means), then player_ship_class is initialized to -1 as well.
```
I'm not sure about this condition. The original condition was true if the ship name couldn't be found but here the condition is only true if the ship name wasn't found and `ship_class` was `-1`.

MageKing17 replied:
The `if ( ship_class == -1)` conditional was there before. The logic is equivalent.

asarium replied:
Ohh, that was folded away. Then this is not an issue...

asarium commented on code/parse/lua.cpp:
```diff
@@ -5618,7 +5618,7 @@ ADE_FUNC(__tostring, l_Shipclass, NULL, "Ship class name", "string", "Ship class
 		return ade_set_error(L, "s", "");
 
-	if(idx < 0 || idx > Num_ship_classes)
+	if(idx < 0 || idx >= static_cast<int>(Ship_info.size()))
```
I think the new version is correct but  the whole time these conditions were wrong? Weird that this never caused issues.

asarium commented on code/parse/sexp.cpp:
```diff
@@ -7544,7 +7546,7 @@ int sexp_get_damage_caused(int node)
 	}
 	
-	Assert ((ship_class > -1) && (ship_class < MAX_SHIP_CLASSES));
+	Assertion((ship_class > -1) && (ship_class < static_cast<int>(Ship_info.size())), "Invalid ship class '%d' passed to sexp_get_damage_caused() (should be >= 0 and < %d); get a coder!\n", ship_class, Ship_info.size());
```
The `Ship_info.size()` is missing a cast to int.

asarium commented on code/pilotfile/csg.cpp:
```diff
@@ -184,10 +184,10 @@ void pilotfile::csg_write_info()
 
 	// ship list
-	cfwrite_int(Num_ship_classes, cfp);
+	cfwrite_int(Ship_info.size(), cfp);
```
Missing cast to `int`.

z64555 commented on code/io/keycontrol.cpp:
```diff
@@ -490,11 +490,11 @@ void debug_cycle_player_ship(int delta)
 	while ( TRUE ) {
 		si_index += delta;
-		if ( si_index > Num_ship_classes ){
+		if ( si_index >= static_cast<int>(Ship_info.size()) ){
```
Is this >= correct?

MageKing17 replied:
Well, if `si_index == Ship_info.size()`, then the index access on line 499 below (`sip = &Ship_info[si_index];`) will be out-of-bounds.

asarium commented on code/ship/ship.cpp:
```diff
@@ -660,373 +658,1053 @@ int lightningtype_match(char *p)
+	glowpoint_bank_override_map = other.glowpoint_bank_override_map;
+}
+
+void ship_info::move(ship_info& other)
```
This should probably use `ship_info&&` to make sure move semantics are properly used in the function.

MageKing17 replied:
As a helper function called by the move-constructor and move-operator, using ship_info&& here would be an error...

asarium replied:
You have to use std::move to convert it back into a rvalue reference.

asarium commented on code/ship/ship.cpp:
```diff
@@ -4125,34 +4885,34 @@ int get_default_player_ship_index()
 	if(strlen(default_player_ship))
 		return;
 
+	auto it = Ship_info.cbegin();
```
Why isn't this declared within the `for` loop?

MageKing17 replied:
So it can be reused in the for loop below.

asarium replied:
But there it gets set to cbegin() which overwrites the previous value so this only adds more variables to the scope which causes confusion.
IMO it would be better to declare two iterator variables in the respective for loops.

MageKing17 replied:
As far as I am aware, it's preferable to declare a variable once and then reuse it (see also: every single function where "int i;" is declared at the beginning and then "i" is used as a loop iterator multiple times in a row, being reset each time). Is it actually preferable to declare a new variable every time we have a for loop? Or will the compiler optimize both so that the difference is irrelevant and it all comes down to which source reads better?

asarium replied:
Every compiler should be able to optimize the code. The "declare once and reuse" rule may be accurate for C but it certainly isn't for C++.

chief1983 replied:
Ok I see this was already rectified but I thought I could clarify here.  The old single 'int i' syntax was done mostly to handle Visual C++ 6, as it does not consider for loops a separate scope, so two for loops each declaring 'int i' would lead to duplicate declaration of i.  It's the only compiler I'm aware of that ever had that issue that we've supported, and if we have indeed officially dropped any and all VC6 support then we can stop catering to that behavior.  So in-loop iterator declaration should be fine, and it should probably be the preferred way going forward.

asarium commented:
I went through all the changes and added notes about things that I noticed.

Now there is another thing I noticed. You added a copy constructor and a assignment operator to ship_class. I think it would be best to not have these (make them private and not implemented) to make sure no one accidentally copies a ship_info instance. The vector will use the move operator (we'll have to confirm if VS2010 properly supports that) and copying a ship_info instance could be done explicitly via the `clone()` function.

MageKing17 commented:
Literally the whole point of the copy-assignment operator was so that ship templates could be implemented using it, but I suppose clone() could be made public instead, if it improves vector performance...

asarium commented:
I know but it's way too easy to accidentally copy an instance if there is a copy constructor. The clone approach would make sure that the instance is only cloned if it really should be.

z64555 commented on code/ship/ship.cpp:
```diff
@@ -1417,25 +2139,36 @@ int parse_and_add_briefing_icon_info()
 		destpos = sip->short_name;
 		srcend = srcpos + strlen(sip->name);
-		while(srcpos < srcend)
+		while(srcpos <= srcend)
```
Is this right?

MageKing17 replied:
Yes.

asarium commented:
Apparently VS2010 accepts the move constructor and operator. I'll give this another test on linux to see if there are remaining issues.

asarium commented:
After merging master into my local branch this seems to be working well although I haven't tested the template functionality.
I think it would be better to wait until #274 is merged so any new warnings generated by these changes (if any) can be easily identified and fixed.

chief1983 commented:
Found some warnings not in master (after perusing the entire debug build output twice to compare the two, need a better way to do this).

CompileC build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/autopilot.o /Users/cliff.gordon/fs2_open_git/code/autopilot/autopilot.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
    cd /Users/cliff.gordon/fs2_open_git/projects/Xcode4
    export LANG=en_US.US-ASCII
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c++ -arch x86_64 -fmessage-length=214 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -fcolor-diagnostics -std=c++11 -stdlib=libc++ -Wno-trigraphs -fpascal-strings -O0 -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-non-virtual-dtor -Wno-overloaded-virtual -Wno-exit-time-destructors -Wno-missing-braces -Wparentheses -Wswitch -Wunused-function -Wno-unused-label -Wno-unused-parameter -Wunused-variable -Wunused-value -Wempty-body -Wuninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wconstant-conversion -Wint-conversion -Wbool-conversion -Wenum-conversion -Wshorten-64-to-32 -Wno-newline-eof -Wno-c++11-extensions -D_DEBUG -DSCP_UNIX -DUSE_OPENAL -DAPPLE_APP -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk -fasm-blocks -fstrict-aliasing -Wdeprecated-declarations -Winvalid-offsetof -mmacosx-version-min=10.7 -g -Wno-sign-conversion -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/code.hmap -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/Debug/include -I../../code -IFrameworks/Ogg.framework/Headers -IFrameworks/Vorbis.framework/Headers -IFrameworks/Theora.framework/Headers -IFrameworks/SDL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/System/Library/Frameworks/OpenAL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/System/Library/Frameworks/OpenGL.framework/Headers -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/DerivedSources/x86_64 -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/DerivedSources -Wno-write-strings -Wno-char-subscripts -Wno-shorten-64-to-32 -F/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/Debug -FFrameworks -MMD -MT dependencies -MF /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/autopilot.d --serialize-diagnostics /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/autopilot.dia -c /Users/cliff.gordon/fs2_open_git/code/autopilot/autopilot.cpp -o /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/autopilot.o
/Users/cliff.gordon/fs2_open_git/code/autopilot/autopilot.cpp:248:47: warning: format specifies type 'int' but the argument has type 'size_type' (aka 'unsigned long') [-Wformat]
                                shipp->ship_name, shipp->ship_info_index, Ship_info.size());
                                                                          ^~~~~~~~~~~~~~~~
In file included from /Users/cliff.gordon/fs2_open_git/code/autopilot/autopilot.cpp:8:
In file included from ../../code/ai/ai.h:15:
In file included from ../../code/ai/ai_profiles.h:13:
../../code/globalincs/pstypes.h:321:48: note: expanded from macro 'Assertion'
                                WinAssert(#expr,__FILE__,__LINE__, msg , ##__VA_ARGS__ );\
                                                                           ^
1 warning generated.



CompileC build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/ship.o /Users/cliff.gordon/fs2_open_git/code/ship/ship.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
    cd /Users/cliff.gordon/fs2_open_git/projects/Xcode4
    export LANG=en_US.US-ASCII
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c++ -arch x86_64 -fmessage-length=214 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -fcolor-diagnostics -std=c++11 -stdlib=libc++ -Wno-trigraphs -fpascal-strings -O0 -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-non-virtual-dtor -Wno-overloaded-virtual -Wno-exit-time-destructors -Wno-missing-braces -Wparentheses -Wswitch -Wunused-function -Wno-unused-label -Wno-unused-parameter -Wunused-variable -Wunused-value -Wempty-body -Wuninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wconstant-conversion -Wint-conversion -Wbool-conversion -Wenum-conversion -Wshorten-64-to-32 -Wno-newline-eof -Wno-c++11-extensions -D_DEBUG -DSCP_UNIX -DUSE_OPENAL -DAPPLE_APP -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk -fasm-blocks -fstrict-aliasing -Wdeprecated-declarations -Winvalid-offsetof -mmacosx-version-min=10.7 -g -Wno-sign-conversion -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/code.hmap -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/Debug/include -I../../code -IFrameworks/Ogg.framework/Headers -IFrameworks/Vorbis.framework/Headers -IFrameworks/Theora.framework/Headers -IFrameworks/SDL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/System/Library/Frameworks/OpenAL.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.10.sdk/System/Library/Frameworks/OpenGL.framework/Headers -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/DerivedSources/x86_64 -I/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/DerivedSources -Wno-write-strings -Wno-char-subscripts -Wno-shorten-64-to-32 -F/Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/Debug -FFrameworks -MMD -MT dependencies -MF /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/ship.d --serialize-diagnostics /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/ship.dia -c /Users/cliff.gordon/fs2_open_git/code/ship/ship.cpp -o /Users/cliff.gordon/fs2_open_git/projects/Xcode4/build/FS2_Open.build/Debug/code.build/Objects-normal/x86_64/ship.o
/Users/cliff.gordon/fs2_open_git/code/ship/ship.cpp:1747:228: warning: format specifies type 'int' but the argument has type 'size_type' (aka 'unsigned long') [-Wformat]
  ...before '%s'; maximum is %d, so only the first %d will be used\nPlease check also the debug log as it may contain other ship classes which are over the limit", buf, MAX_SHIP_CLASSES, Ship_info.size());
                                                   ~~                                                                                                                                      ^~~~~~~~~~~~~~~~
                                                   %lu
/Users/cliff.gordon/fs2_open_git/code/ship/ship.cpp:13997:47: warning: format specifies type 'int' but the argument has type 'size_type' (aka 'unsigned long') [-Wformat]
                                shipp->ship_name, shipp->ship_info_index, Ship_info.size());
                                                                          ^~~~~~~~~~~~~~~~
In file included from /Users/cliff.gordon/fs2_open_git/code/ship/ship.cpp:15:
In file included from ../../code/ai/aigoals.h:16:
../../code/globalincs/pstypes.h:321:48: note: expanded from macro 'Assertion'
                                WinAssert(#expr,__FILE__,__LINE__, msg , ##__VA_ARGS__ );\
                                                                           ^
2 warnings generated.

chief1983 commented:
Ok, I don't think this is going to get any safer in a PR, so we need to get this in a nightly now.

[![LGTM](http://lgtm.in/p/yDP32iPIn)](http://lgtm.in/i/yDP32iPIn)
